### ISSUE TITLE
use wavernn to make TTS

### ISSUE THREAD

**Issue by maozhiqiang** [fatchord/WaveRNN#1]
Can we use wavernn to make TTS, what is the input, thank you

**Comment by fatchord** [comment]
Hi there, I believe you can, in the paper that's what they said they did with it. They are using linguistic/pitch features as conditioning to wavernn. 

Right now I'm trying to condition with upsampled mel spectrograms (like the tacotron2 vocoder) but the sound quality isn't great - I'm getting a whispering/gravelly sound, like the speaker has a really bad cold! I still have a couple more ideas to test out so if I have any success I'll upload it to the repo.

**Comment by maozhiqiang** [comment]
@fatchord Thank you for your quick reply，There are some points in the paper that I haven't fully understood. So I'm using your code to understand it！

**Comment by fatchord** [comment]
@maozhiqiang You're very welcome. Please keep in mind that there are some details (optimizer, learning rate, conditioning equations, signal splitting technique etc etc) that were left out of the paper so I've had to basically guess at what these are and improvise accordingly.

**Comment by maozhiqiang** [comment]
@fatchord  Thank you for your detailed explanation！

**Comment by fatchord** [comment]
Hi there, I'm not 100% sure but the matrix multiplication for the inputs is very small compared to the others in the equations so perhaps that's why it's left out of N.

As for conditioning, I've tried a couple of things - I've passed the conditioning features through a dense layer and then biased the gates. I've also tried concatenating them to the input. Both are working pretty much the same but the results are only okayish - the speech is intelligible but it has this trembling quality which might point to my upsampling network being incorrect . Have a listen: 
[wavernn_conditioned.tar.gz](https://github.com/fatchord/WaveRNN/files/1945874/wavernn_conditioned.tar.gz)

I'm currently training another model with different upsampling network & bigger dataset. If results are good I'll upload to the repo.

**Comment by lifeiteng** [comment]
@fatchord very nice result, does them use `ground-truth c_t`?
The conditioning formula is missed in the paper, maybe we should email with the authors.

I concatenate Linguistic(green line) / Acoustic (red line) features with `x_t`, training on [cmu_us_slt_arctic](http://www.speech.cs.cmu.edu/cmu_arctic/cmu_us_slt_arctic/):
![2018-05-08 19 28 56](https://user-images.githubusercontent.com/3107839/39755050-f2c6225a-52f6-11e8-935d-5e10813d082a.png)

**Comment by fatchord** [comment]
@lifeiteng thanks, it's not a terrible result, at least it's working somewhat. However on the same dataset I get a smoother quality with wavenet(8bit): 
[wavenet_gen_glr_dataset.tar.gz](https://github.com/fatchord/WaveRNN/files/1986913/wavenet_gen_glr_dataset.tar.gz)

I'm not sure what you mean by `ground-truth c_t` - do you mean conditioning features at time t?

By the way, I'd love to hear your wavernn samples so feel free to post them here!

**Comment by lifeiteng** [comment]
@fatchord I mean `Are those samples synthesized only on Acoustic Features, not include ground-truth coarse & fine as inputs(like training)`.

Doing experiments now.

**Comment by lifeiteng** [comment]
BTW, you can try 9-bit or 10-bit u-law in `WaveNet` .

**Comment by maozhiqiang** [comment]
@fatchord hi! for mel as a local condition , do you  use upsample ! how to as input of wavernn!

**Comment by fatchord** [comment]
@lifeiteng Re:generation - thanks for clarifying - those are synthesized with mel spectrograms only, the model didn't see any ground truth audio samples when generating.

Re: 9 or 10 bit wavenet - actually I've been wanting to try this out but my gpu has been busy with other experiments for a while now. Did you try it? Any success with it?

**Comment by lifeiteng** [comment]
@fatchord wavenet - I try 9-bit (e.g. 512 softmax) in wavenet, got much more clean result than 8-bit version. I saw a paper which claim 10-bit is better than 8-bit recently.

Now, I like wavernn more, because the training of wavenet needs several days GPU time to get reasonable result.

**Comment by maozhiqiang** [comment]
@fatchord thanks ! i well implement this like you said!

**Comment by fatchord** [comment]
@lifeiteng Sounds very interesting, have you got a link to the paper?

**Comment by lifeiteng** [comment]
@fatchord  http://festvox.org/blizzard/bc2017/USTC-NELSLIP_Blizzard2017.pdf 
```
As the original 8bit quantization introduced quantization
noise in synthetic speeches, we proposed to use a 10bit quantization
scheme instead, in order to alleviate this problem.
WaveNet with 3 blocks, which was 30 layers in total, was used
in our system.
```

**Comment by fatchord** [comment]
@lifeiteng Thanks! Nice read, I really love the idea of using a GAN to fix the smoothness from the mse loss.

**Comment by shreyasnivas** [comment]
Hi @fatchord - you've got some nice quality results there both from wavernn and wavenet!

Did you manage to overcome that trembling quality with your wavernn 198k sample? it's something we've been puzzling over too...

Our intuition points to the following areas 
1. The continuity of the RNN is broken by re-initializing the hidden state after 960 samples (and would point to a periodic tremble)
2. The upsampling network may produce artifacts
3. If all else fails, more data..?

To preserve continuity, we are looking into TBPTT and longer input sequence lengths (like 2000 instead of 960) as well as increasing the length of our batches from the dataloader so we can carry forward the hidden states longer on each step.
 
Our upsampling is the same conv2d approach as in Kay Bayashi's wavenet. (we tried different stride/conv2d settings, like stacking smaller upsampling convs etc but found that the single 2d conv that upsamples the mel_spectrograms 'hop_length' times works best).

The last appraoch is to train it on a huge dataset like LJSpeech and just wait it out which is where we are now, but our initial overfitting attempts on wavernn didn't seem to get rid of the jittery/trembling quality... 

Did you find anything that worked? i.e wavernn comparable to wavenet quality?

cheers,
shreyas

**Comment by MlWoo** [comment]
@lifeiteng  Did u message the authors to make sense the conditioning formula?

**Comment by lifeiteng** [comment]
@MlWoo yes, but not got the response.

**Comment by lifeiteng** [comment]
That's a data feeding bug. Actually, I'm not lucky on wavernn.

**Comment by MlWoo** [comment]
what a pity!

**Comment by jjj8080** [comment]
@fatchord Is there any chance you could upload the code version you used to condition on mel-spectrograms?

**Comment by tiberiu44** [comment]
Hi guys,

I've managed to implement a modified version of WaveRNN (I use two parallel RNN layers instead of modifying the cell) and slightly modified version of Tacotron to output conditioned spectograms. I tested the approach on Romanian (directly on characters) and I got some pretty ok results. You can check them out here (I also added e results obtained with HTS trained on the same corpus with lots of features extracted from text - syllables, phonetic transcription, POS, chunking etc.):

https://tiberiu44.github.io/TTS-Cube/ 

the project's repo is
https://github.com/tiberiu44/TTS-Cube

I'm currently working on adding some more text-features (unsupervised) and on conditioning the model on multiple speakers. 

Hope this helps,
Tibi

**Comment by MlWoo** [comment]
@tiberiu44 Thanks a lot for your contribution. I have tried to train the vocoder with your code, But it is very slow to train the vocoder on 1080ti. I want to confirm the info because I am not familar to dynet framework.

**Comment by MlWoo** [comment]
@tiberiu44  I did compile the dynet useing cuda backend.  I have not checked you model yet, maybe the model is large.  Thanks for your reply.
